### PR TITLE
Transaction Helper Functions

### PR DESCRIPTION
**Pull request by StevenWadeOddball**
# EASI-3507

## Changes and Description

- Implement a WithFunction to Wrap Transactions
- Create an interface to be able to pass either a store or a transaction.
   - Implement these for create Model Plan

## How to test this change
1. Create a model plan
2. Verify that there aren't any errors

Code to pay close attention to
1. [With Transactions](https://github.com/CMSgov/mint-app/compare/main...EASI-3507/db_transaction_helpers_withFunction#diff-92b79a5814d551bc30fd30f4e54e699ff2eff47d5f153be09e1180092e952a0f)
   a. and related test file
      i. Note, this could also go in the resolvers package so it has db clean methods etc.
2. [New Model Plan Resolver](pkg/graph/resolvers/model_plan.go)
3. [Named Preparer interface](pkg/storage/named_preparer.go)

## PR Author Review Checklist

- [x] Met the ticket's acceptance criteria, or will meet them in a subsequent PR.
- [x] Added or updated tests for backend resolvers or other functions as needed.
- [x] Added or updated client tests for new components, parent components, functions, or e2e tests as necessary.
- [x] Updated the [Postman Collection](../MINT.postman_collection.json) if necessary.

## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.

### CODE REVIEW

File: cmd/dbseed/resolver_wrappers.go [hunk 1]
```diff
@@ -112,13 +112,14 @@ func (s *Seeder) addPlanCollaborator(
 
 	collaborator, _, err := resolvers.CreatePlanCollaborator(
 		context.Background(),
+		s.Config.Store,
+		s.Config.Store,
```
Comment (OddTomBrooks) [inline cmd/dbseed/resolver_wrappers.go:116]: Is this intended to be added twice?

File: pkg/graph/resolvers/added_as_collaborator_email_test.go [hunk 2]
```diff
@@ -62,13 +62,14 @@ func (s *ResolverSuite) TestAddedAsCollaboratorEmail() {
 
 	_, _, err := CreatePlanCollaborator(
 		context.Background(),
+		s.testConfigs.Store,
+		s.testConfigs.Store,
```
Comment (OddTomBrooks) [inline pkg/graph/resolvers/added_as_collaborator_email_test.go:66]: Like above, why are we passing the store twice?

File: pkg/graph/resolvers/plan_collaborator.go [hunk 3]
```diff
@@ -72,6 +75,61 @@ func CreatePlanCollaborator(
...
+// 	if emailService != nil && emailTemplateService != nil {
+// 		err = sendCollaboratorAddedEmail(emailService, emailTemplateService, addressBook, collabAccount.Email, modelPlan)
+// 		if err != nil {
+// 			return retCollaborator, planFavorite, err
+// 		}
+// 	}
+
+// 	return retCollaborator, planFavorite, nil
+// }
+
```
Comment (OddTomBrooks) [inline pkg/graph/resolvers/plan_collaborator.go:132]: Should this be removed?

File: pkg/graph/resolvers/plan_collaborator_test.go [hunk 4]
```diff
@@ -69,13 +69,14 @@ func (suite *ResolverSuite) TestCreatePlanCollaborator() {
 
 	collaborator, _, err := CreatePlanCollaborator(
 		context.Background(),
+		suite.testConfigs.Store,
+		suite.testConfigs.Store,
```
Comment (OddTomBrooks) [inline pkg/graph/resolvers/plan_collaborator_test.go:73]: Same concern as above

File: pkg/storage/model_planStore.go [hunk 5]
```diff
@@ -68,18 +68,19 @@ func (s *Store) ModelPlanGetByModelPlanIDLOADER(_ *zap.Logger, paramTableJSON st
...
 	}
-	stmt, err := s.db.PrepareNamed(modelPlanCreateSQL)
+
+	stmt, err := np.PrepareNamed(modelPlanCreateSQL)
 	if err != nil {
 		logger.Error(
 			fmt.Sprintf("Failed to create model plan with error %s", err),
 			zap.String("user", plan.CreatedBy.String()),
 		)
+		// t.errors = append(t.errors, err) // TODO: SW should we just return the error? It gets appended in the parent transaction
```
Comment (OddTomBrooks) [inline pkg/storage/model_planStore.go:83]: Good question, @ClayBenson94 what do you think? This has been a question before.

File: pkg/storage/model_planStore.go [hunk 6]
```diff
@@ -91,6 +92,7 @@ func (s *Store) ModelPlanCreate(logger *zap.Logger, plan *models.ModelPlan) (*mo
 
 	err = stmt.Get(&retPlan, plan)
 	if err != nil {
+		// t.errors = append(t.errors, err) // TODO: SW should we just return the error? It gets appended in the parent transaction
```
Comment (OddTomBrooks) [inline pkg/storage/model_planStore.go:95]: Ditto above

**Review by OddTomBrooks — COMMENTED** [review]
The biggest code smell is passing the store twice. It appears that the store is acting as an interface. I would prefer passing an adapter which accepts the store and wraps the behavior so that we contain and isolate some of the tech debt, but this is not the end of the world. It's just something that might confuse new developers. Overall, great job as usual Steven!

File: pkg/storage/model_planStore.go (hunk 5 above)
Comment (StevenWadeOddball) [reply pkg/storage/model_planStore.go:83]: sorry, this should be removed, just forgot to

File: pkg/graph/resolvers/added_as_collaborator_email_test.go (hunk 2 above)
Comment (StevenWadeOddball) [reply pkg/graph/resolvers/added_as_collaborator_email_test.go:66]: yes, see open conversation

File: cmd/dbseed/resolver_wrappers.go (hunk 1 above)
Comment (StevenWadeOddball) [reply cmd/dbseed/resolver_wrappers.go:116]: [EASI-3657](https://jiraent.cms.gov/browse/EASI-3657) Is the ticket I refer to that will resolve this.

**Pull request merged**